### ISSUE TITLE
Rundeck scheduling and dependency enhancements

### ISSUE THREAD

**Issue by Daryes** [rundeck/rundeck#2703]
Issue type: Enhancement Request

Hello,
I've tried to find a complete scheduler, and found rundeck recently, being the closest to what I was seeking after working for a long time on some business schedulers (maybe too much time for my sanity).
Here's a list of proposed additions, most of them related to 2 items (schedules and dependencies) that will help a lot for more complexes uses.
Some of them might be obscure initially, but are related to another right after.

It would be nice to have those, if not all, implemented in Rundeck.

1. **separate job definition and job scheduling**

This will allow one shot / automated scheduling, with different settings and values of any variables declared in the definition
The scheduling also need to allow to be changed through rundeck-cli in the command line
This might require an additional panel to manage the schedules

2. **scheduled queue**

When a job is scheduled or manually launched, it'll create a future launch instance being a copy of the schedule definition, which will be executed later by the scheduler at the planned date/time.
Any change done on such object will be restricted to the instance, and not the scheduling model.
This will allow to change a planned launch without altering the settings.

For reducing the workload, launch instance will be calculated 24h before their expected launch date/time, configurable in the general settings

3. **job execution states**

In addition to the running, succeeded, failed and aborted states, jobs will need the planned and waiting states.
Planned is obvious, waiting is a job which reached the expected start time, but cannot yet be launched due to dependencies.

4. **execution date (processing date) in the scheduling queue.**

This parameter will store the planned date the job is scheduled at, with options to force a different one (like d-1 if set to a time before 06h am)

For example, multiple jobs are planned right after the end of working hours. If, for some reason one of them cannot starts before 00h or is still working after this time, the current date will be day+1. Anything using this date won't be able to work correctly without handling this.
Same problem if the day is at the end of the month, or year.

With the (planned) execution date, jobs have in the environment a variable with the scheduled day, allowing them to work as expected 
For the sake of usability, make it 2 variables : one with the locale format, another using the yyyymmdd format (might be with or without '/' )

this information must also be available in the execution history (log and cli), and the activity panel

5. **job dependencies - waiting for**

Dependencies need to be able to wait for other jobs, using the execution date pulled from jobs history to ensure the correct date flow is used.
The waiting time will be configurable, with the option to force the launch at the end. Also, a job started but in the waiting state must also have the option to be immediately forced to run, ignoring the dependencies.

Some scheduler enforce the flow definition, and prevent the possibility of dependencies with a job from a different flow. If implemented, this should be configurable at the global and project level.

As a bonus, the option to have either a hard dependency, waiting for the expected job state, or a soft dependency, only waiting if the job was launched before, either running, completed, failed, or in a waiting state.

6. **job dependencies - dummy event**

Sometimes a job must not or can't be launched. It is then useful to instead create a dummy event of a job with a specified status, that will allow the dependencies to resolve instead of stopping the flow

---

Other possible options : 

7. **more complex job conditional formulae.**

Something along this :
	A=cond 1
	B=cond 2
	...
	formula=(A OR B) AND C

8. **calendar definition**

First usage is to handle workdays and holidays. Starting with a single calendar, but allowing to create and use more. They have to be selectable in the scheduling definitions.

Some scheduler don't allow to select a day in the scheduling, but only through the calendar definitions. Might be more usable to use the calendar as a template of available choice, and keep the day selection in the scheduling.
This will need a dedicated panel anyway.

9. **job execution - retry at step level**

There's currently the option to relaunch a failed job, but not to (re) start a job at a step level.

10. **flow presentation**

The flow is the list of the jobs scheduled for a given date (processing date)
I'm not a personal fan of this, but it has it's use

11. **flow corridor execution**

It's a sublevel for job scheduling, allowing to set for a job a different time scheduling, and different variables values, with default being the first corridor.
For example, when you have the same job you want to execute on multiple sub environments, with the only differences between them being the servers names, and values for the job variables.

**Comment by gschueler** [comment]
this is a great feature rundown for scheduling, thank you!

**Comment by padraiglennon** [comment]
There are some excellent suggestions above and if Rundeck were able to implement even very simplistic solutions to most of the suggestions I feel it would provide an avenue for companies to move away from old job schedulers which are very expensive to Rundeck..

**Comment by lfpalacios** [comment]
Any updates on this?

**Comment by Daryes** [comment]
For those keeping an eye on this, I've built a Linux-only plugin working with Rundeck's API to handle the point **5) job dependencies - waiting for**, and partially the **6) job dependencies - dummy event**  

A bonus (or not) is the simulation of a complete job's flow from day+0/14h00 to day+01/13h59, that was required to create a limit on the search depth for a previous launch of a job.  
I'm not much a fan of flows with a hard time limit, but until Rundeck provides the execution date information on each launch, there's not many possibilities left. 

I'm running this since something like a year with many daily jobs linked together through this plugin, no much of a problem left, aside the limitation to not able to simulate an empty launch when a job stops the remaining flow after an error occurred.

Available on my repo : https://github.com/Daryes/rundeck-dependencies-plugin

**Comment by lightarrow** [comment]
yes this is still relevant and something we are looking at in enterprise tools.

**Comment by yogeshkumark** [comment]
what is waiting in the release this enhancement ?

**Comment by lightarrow** [comment]
I have been able to resolve advanced scheduling issues with API calls, or the RD command line tools.

**Comment by am312** [comment]
8 calendar definition  <-- NEED THIS

**Comment by otterz** [comment]
Any updates on this? I'd like to see feature 1-4 :-)

**Comment by mungo312** [comment]
Features 1, 2, 3, 4 would be really great.

**Comment by lightarrow** [comment]
With job rulesets most of this feature request is already complete. 
I would not want most of the features implemented in a complex workflow. 

1. separate job definition and job scheduling 
This is feature complete with project schedules. 

2. Project calendars and schedules fulfills this enhancement request. 
3. Job rulesets and context fulfills this request.
4. Using calendars with blackout dates and schedule dated fulfills this request. 
5. Job rulesets and context fulfills this request.
6. job step context could be extended to make the contexts visible within the GUI to schedulers. 
Job context can be one of depends on, waiting for, blacked out due to time, missed run due to windows. 

The rest are covered quite well 
10. Job flows and forecasting needs a lightweight answer. 
A calendar based job flow makes the most sense to me. 
To put it simple. If a job is going to run on a day and time put it in a list in a calendar view for that day. 

ok these features are all well and good but retry's when processing account balances would be a nightmare. 
in most cases you need idempotent batch jobs to even think about retries. 
most Bach jobs involve reconciliation of disparate data between two organizations. 
This causes a missed run to have impact to customers.
it also causes a retry to have a bit of risk based on your batch jobs age, state, and ability to provide atomic transactions.

**Comment by Daryes** [comment]
Just to chime about job's retries : it is not very hard, as long as your data has a date/time information.

**Comment by Daryes** [comment]
Sorry, closing the issue was an error.

**Comment by lightarrow** [comment]
@Daryes I agree, but sometimes the job you run are legacy and will not get updated, and are on completely different systems, containers, DB's ect.... very risky.

**Comment by fdevans** [comment]
Thanks for the report. As part of a cleanup of our issue tracker, we're closing issues older than 12 months, in line with Rundeck's support policy of one year from a release's date (see the [release calendar](https://docs.rundeck.com/docs/history/release-calendar.html) for currently supported versions). This isn't a judgment that the report is invalid — many of these are likely still real — but we can't verify against a version this old.

If you're still seeing this on a currently supported release (6.1.x or 5.20.x+), please open a new issue using our [bug report template](https://github.com/rundeck/rundeck/issues/new?template=bug_report.yml) with your version and reproduction steps. Please don't reopen this issue — a fresh issue keeps things tracked against a supported version and gets it in front of us properly. We're actively working on improving how issues get submitted and triaged going forward.

Thanks for your patience,
The Rundeck Team
